src/dotenv/main.py

```diff
@@ -151,13 +151,16 @@ def set_key(dotenv_path, key_to_set, value_to_set, quote_mode="always", export=F
     If the .env path given doesn't exist, fails instead of risking creating
     an orphan .env somewhere in the filesystem
     """
-    value_to_set = value_to_set.strip("'").strip('"')
+    if quote_mode not in ("always", "auto", "never"):
+        raise ValueError("Unknown quote_mode: {}".format(quote_mode))
 
-    if " " in value_to_set:
-        quote_mode = "always"
+    quote = (
+        quote_mode == "always"
+        or (quote_mode == "auto" and not value_to_set.isalnum())
+    )
 
-    if quote_mode == "always":
-        value_out = '"{}"'.format(value_to_set.replace('"', '\\"'))
+    if quote:
+        value_out = "'{}'".format(value_to_set.replace("'", "\\'"))
```

Comment:
Just so you know, this is not the correct way to escape strings surrounded by single quotes when it comes to shells like Bash ([this is the way](https://stackoverflow.com/a/26165123/247696)).

I don't know if this project has a goal of making `.env` files compatible with Bash, if it does, than this code will fail in this edge-case. These commands in the Bash command-line will demonstrate my point:

```
$ printf %s 'one'"'"'two'
one'two
$ printf %s 'one\'two'
> 
```

(The last command doesn't run after pressing enter, but waits for a terminating single quote.)

Reply:
That's a very good point. I had actually forgotten about that rule.  I'll see if we can do better with regards to matching Bash's behavior but I'm not sure it's worth it.

Reply:
@bbc2 I believe we are facing this bug. We have single quotes inside value, it is being [wrongly] escaped as `\'` (which is not the right way) and `source .env` of resulting file fails